girder/api/sftp.py

```diff
@@ -0,0 +1,302 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+###############################################################################
+#  Copyright 2016 Kitware Inc.
+#
+#  Licensed under the Apache License, Version 2.0 ( the "License" );
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+###############################################################################
+
+import os
+import paramiko
+import six
+import stat
+import time
+
+from girder import logger
+from girder.models.model_base import AccessException, ValidationException
+from girder.utility.path import lookUpPath, NotFoundException
+from girder.utility.model_importer import ModelImporter
+from six.moves import socketserver
+
+DEFAULT_PORT = 8022
+
+
+def _handleErrors(fun):
+    @six.wraps(fun)
+    def wrapped(*args, **kwargs):
+        try:
+            return fun(*args, **kwargs)
+        except NotFoundException:
+            return paramiko.SFTP_NO_SUCH_FILE
+        except ValidationException:
+            return paramiko.SFTP_FAILURE
+        except AccessException:
+            return paramiko.SFTP_PERMISSION_DENIED
+        except Exception:
+            logger.exception('SFTP server internal error')
+
+    return wrapped
+
+
+def _getFileSize(file):
+    if file.get('assetstoreId'):
+        return file['size']
+    elif file.get('linkUrl'):
+        return len(file['linkUrl'])
+
+
+class _FileHandle(paramiko.SFTPHandle, ModelImporter):
+    def __init__(self, file):
+        """
+        Create a file-like object representing a file blob stored in Girder.
+
+        :param file: The file object being opened.
+        :type file: dict
+        """
+        super(_FileHandle, self).__init__()
+
+        self.file = file
+
+    def read(self, offset, length):
+        stream = self.model('file').download(
+            self.file, headers=False, offset=offset, endByte=offset + length)
+        return b''.join(stream())
```

Comment:
Yup, I looked for another streaming API, but couldn't see one. A sensible limit is a good idea.